### PR TITLE
Add Error Handler For POST EndPoint-3

### PR DESCRIPTION
**Pull request by DennisShelton04**
### **PR Title:**
- **"Add Common Error Handler with `BookmarkManagerException` Class"**

#### **Overview:**
This PR introduces a common error handling mechanism by adding the `BookmarkManagerException` class, which extends `RuntimeException` and standardises error responses across the application.

#### **New Feature:**
1. **`BookmarkManagerException` Class:**
   - A custom exception class that holds a list of error details in a standardized format.
   - The constructor accepts:
     - `message`: A descriptive error message.
     - `errors`: A list of errors, each represented as a `Map<String, String>`, detailing specific issues.

2. **`ErrorResponse` Record:**
   - This record is used to wrap error responses with two fields:
     - `message`: The general error message.
     - `errors`: A list of specific error details (maps containing error fields, messages, and error codes).

### CODE REVIEW

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 1]
```diff
@@ -44,6 +40,40 @@ public Bookmark addBookmark(Bookmark bookmark) {
     return bookmarkRepository.save(bookmark);
   }
 
+  private void precheckBookmark(Bookmark bookmark) {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]:
```suggestion
  private void validateBookmark(Bookmark bookmark) {
```

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 2]
```diff
@@ -1,36 +1,32 @@
...
-    // Validate the URL and title (basic example)
-    if (bookmark.getUrl() == null || bookmark.getUrl().isBlank()) {
-      throw new IllegalArgumentException("URL cannot be blank");
-    }
-    if (bookmark.getTitle() == null || bookmark.getTitle().isBlank()) {
-      throw new IllegalArgumentException("Title cannot be blank");
-    }
 
-    // If folderId is provided, associate the bookmark with the folder
+    precheckBookmark(bookmark);
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]:
```suggestion
    validateBookmark(bookmark);
```

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 3]
```diff
@@ -1,36 +1,32 @@
...
 @RequiredArgsConstructor
 public class BookmarkService {
+  public static final String FIELD = "field";
+  public static final String MESSAGE = "message";
+  public static final String ERROR_CODE = "errorCode";
   private final BookmarkRepository bookmarkRepository;
   private final FolderRepository folderRepository;
 
 
   public Bookmark addBookmark(Bookmark bookmark) {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]:
```suggestion
  public Bookmark createBookmark(Bookmark bookmark) {
```
This operation should be transactional, so use @Transactional

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 1 above)
Comment (Bala-raj) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: This validateBookmark should also be used in the updateBook Method.

**Review by Bala-raj — CHANGES_REQUESTED** [review]

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 3 above)
Comment (Bala-raj) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: Also updateMethod should be transactional.

Comment (DennisShelton04) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: Added those Changes

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 1 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: Added validateBookmark for updateBookmark.

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 2 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: Renamed the precheckBookmark ->  validateBookmark

File: Bookmark-Manager/build.gradle [hunk 4]
```diff
@@ -23,10 +23,13 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
 	implementation 'org.projectlombok:lombok:1.18.26'
+	testImplementation 'junit:junit:4.13.1'
+	testImplementation 'junit:junit:4.13.1'
```
Comment (Bala-raj) [inline Bookmark-Manager/build.gradle]: why twice?

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 5]
```diff
@@ -1,36 +1,30 @@
...
-    if (bookmark.getTitle() == null || bookmark.getTitle().isBlank()) {
-      throw new IllegalArgumentException("Title cannot be blank");
-    }
+  @Transactional
+  public Bookmark createBookmark(Bookmark bookmark) {
 
-    // If folderId is provided, associate the bookmark with the folder
+    validateBookmark(bookmark);
     if (bookmark.getFolderId() != null) {
       Optional<Folder> folderOptional = folderRepository.findById(bookmark.getFolderId());
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: Why do you not include validateBookmark to verify the folder ID on updates and patches?

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 6]
```diff
@@ -44,12 +38,46 @@ public Bookmark addBookmark(Bookmark bookmark) {
...
+    if (bookmark.getUserId() == null) {
+      errors.add(Map.of(
+              FIELD, "userId",
+              MESSAGE, "User ID cannot be null",
+              ERROR_CODE, "INVALID_USER_ID"
+      ));
+    }
+
+    if (!errors.isEmpty()) {
+      throw new BookmarkManagerException("Invalid Arguments",errors);
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: The code may not be formatted correctly, in my opinion. Make sure of the basics, please.

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 7]
```diff
@@ -44,12 +38,46 @@ public Bookmark addBookmark(Bookmark bookmark) {
     return bookmarkRepository.save(bookmark);
   }
 
+  private void validateBookmark(Bookmark bookmark) {
+    List<Map<String, String>> errors = new ArrayList<>();
+
+    if (bookmark.getUrl() == null || bookmark.getUrl().isBlank()) {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java:48]: We are supposed to check whether it is a really valid URL or not, right?

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java [hunk 8]
```diff
@@ -60,7 +88,7 @@ public Bookmark updateBookmark(UUID id, Bookmark bookmarkDetails) {
     if(existingBookmark.getFolderId()!=null) {
       existingBookmark.setFolder(bookmarkDetails.getFolder());
```
Comment (Bala-raj) [inline Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java:112]: We aren't checking if the folder is already available or not.

File: Bookmark-Manager/build.gradle (hunk 4 above)
Comment (Bala-raj) [reply Bookmark-Manager/build.gradle]: And why not JUnit 5?

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+import java.util.UUID;
+import org.junit.jupiter.api.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+class BookmarkServiceTest {
+
+  BookmarkRepository bookmarkRepository = mock(BookmarkRepository.class);
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java:25]: Should use @BeforeEach to set up test fixtures:

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+  FolderRepository folderRepository = mock(FolderRepository.class);
+  BookmarkService bookmarkService = new BookmarkService(bookmarkRepository, folderRepository);
+
+
+
+  @Test
+  void test_create_bookmark_with_valid_data() {
+    UUID folderId = UUID.randomUUID();
+    Bookmark bookmark = new Bookmark();
+    bookmark.setTitle("Sample Title");
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java:35]: Should extract common test data creation into helper methods

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    assertNotNull(savedBookmark);
+    assertEquals("Sample Title", savedBookmark.getTitle());
+    assertEquals("http://example.com", savedBookmark.getUrl());
+  }
+
+    @Test
+    void test_create_bookmark_with_invalid_data() {
+        Bookmark bookmark = new Bookmark();
+
+        try {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Current exception testing could be improved using JUnit 5's assertThrows:

```
BookmarkManagerException exception = assertThrows(BookmarkManagerException.class, 
    () -> bookmarkService.createBookmark(bookmark));
assertEquals("Invalid Arguments", exception.getMessage());

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        bookmark.setUrl("http://example.com");
+        bookmark.setUserId(UUID.randomUUID());
+
+        when(bookmarkRepository.existsById(bookmarkId)).thenReturn(Boolean.TRUE);
+
+        Mockito.doNothing().when(bookmarkRepository).deleteById(bookmarkId);
+
+        bookmarkService.deleteBookmark(bookmarkId);
+
+
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Verification of Mock Interactions
Missing verification of mock interactions in some tests.

```
    verify(bookmarkRepository).deleteById(bookmarkId);
    verify(bookmarkRepository).existsById(bookmarkId);

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+class BookmarkServiceTest {
+
+  BookmarkRepository bookmarkRepository = mock(BookmarkRepository.class);
+  FolderRepository folderRepository = mock(FolderRepository.class);
+  BookmarkService bookmarkService = new BookmarkService(bookmarkRepository, folderRepository);
+
+
+
+  @Test
+  void test_create_bookmark_with_valid_data() {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java:34]: Test Method Naming 
While the current naming is clear, it could be more consistent with common patterns:

// Better:
shouldCreateBookmarkSuccessfully()
shouldThrowExceptionWhenCreatingBookmarkWithInvalidData()

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+    Bookmark savedBookmark = bookmarkService.createBookmark(bookmark);
+
+    assertNotNull(savedBookmark);
+    assertEquals("Sample Title", savedBookmark.getTitle());
+    assertEquals("http://example.com", savedBookmark.getUrl());
+  }
+
+    @Test
+    void test_create_bookmark_with_invalid_data() {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Missing Tests
Consider adding tests for:
- Boundary conditions for URL and title lengths
- Special characters in URLs
- Duplicate bookmark detection (if applicable)
- Case sensitivity handling

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+        try {
+            bookmarkService.getBookmarkById(bookmarkId);
+        } catch (BookmarkNotFoundException e) {
+          assertNotNull(e.getMessage());
+        }
+    }
+
+    @Test
+    void test_update_bookmark() {
```
Comment (Bala-raj) [inline Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Also use "Given-When-Then" style pattern for better readability 

For example:
```
// Given-When-Then style
@Test
void shouldUpdateBookmarkWhenValidIdProvided() {
    // Given
    UUID bookmarkId = UUID.randomUUID();
    Bookmark existingBookmark = new Bookmark();
    existingBookmark.setTitle("Old Title");
    
    Bookmark updatedBookmark = new Bookmark();
    updatedBookmark.setTitle("New Title");
    
    when(bookmarkRepository.findById(bookmarkId)).thenReturn(Optional.of(existingBookmark));
    when(bookmarkRepository.save(any())).thenReturn(updatedBookmark);
    
    // When
    Bookmark result = bookmarkService.updateBookmark(bookmarkId, updatedBookmark);
    
    // Then
    assertEquals("New Title", result.getTitle());
    verify(bookmarkRepository).findById(bookmarkId);
    verify(bookmarkRepository).save(any());
}

**Review by Bala-raj — CHANGES_REQUESTED** [review]

File: Bookmark-Manager/build.gradle (hunk 4 above)
Comment (DennisShelton04) [reply Bookmark-Manager/build.gradle]: I got it from Maven Central. and it has 
https://share.anysnap.app/fss3ImITgyeE
13.1 is stable.
Will update to JUnit5

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 5 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java]: For Folder Id It can optional. For that case I have't added

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java (hunk 12 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Added verify.

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java (hunk 14 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: For Now I have covered basic. In Service Class will add it in upcomming

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java (hunk 15 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: I have seperated  without comment  by line space will add // Given-When-Then styles

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java (hunk 11 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java]: Added Junit 5 also Handle With new Implementation

File: Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java (hunk 10 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/test/java/com/bookmark/bookmarks/BookmarkServiceTest.java:35]: Extracted Test Case with Helper Method

File: Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java (hunk 7 above)
Comment (DennisShelton04) [reply Bookmark-Manager/src/main/java/com/bookmarkmanager/bookmark/BookmarkService.java:48]: Added the check for URL.
